### ISSUE TITLE
N-API Migration

### ISSUE THREAD

**Issue by weagle08** [yocontra/node-gdal-next#7]
Do you have a branch you will be performing the N-API migration on? Would like to throw a couple developers at this to help maybe around the October timeframe and can probably give about 2 months of development time to assist in this. Would like to rethink how the project gets built however because the modifying current gyp files by hand is very error prone and tedious! We have a requirement to operate on Windows so keeping Windows compatibility is a priority for us. We have never done any native wrappers so will be a learning experience/curve as we are also not C/C++ devs. We can definitely help in adding the async javascript/typescript api as well.

**Comment by weagle08** [comment]
You may have already addressed some of my concerns on the build process as we are still on the old node-gdal module by naturalatlas guys.

Thanks,

Ben

**Comment by contra** [comment]
@weagle08 It took me weeks to port everything from the original old repo to this one because there was so much manual work needed - as I was doing it I tried to automate everything as much as possible so some things were improved, but there is still a looot of complexity in the gyp files.

I would absolutely love some help simplifying the current process - I think it will make the migration to N-API easier to first start by finishing the cleanup of what we currently have. I have a basic roadmap here of what I jotted down while I was migrating this project - https://github.com/contra/node-gdal-next/blob/master/ROADMAP.md. I don't have a timeline for the N-API stuff since it isn't urgent now that I am publishing new prebuilds regularly, and haven't made a branch for it yet.

Feel free to take anything off the ROADMAP and send PRs, or file tickets for other cleanup items you notice. I think simplifying the gyp files/build system would be a huge help.

**Comment by mmomtchev** [comment]
What is the current status of the N-API migration? I was planning to add readAsync/writeAsync methods, then I saw this
Are there any proposals for the async API? I was thinking about copying the fs API creating xxxAsync methods that can be promisified

**Comment by contra** [comment]
@mmomtchev Nothing has been started yet by me or anyone else AFAIK - feel free to make a PR if you are interested in implementing it.

For the async APIs I think that pattern works - IIRC returning promises from native extensions is really annoying so callbacks is the path of least friction, and we can promisify in JS.

**Comment by mmomtchev** [comment]
There is also the GDALAsyncReader API which allows for creating a ReadStream-like interface where a caller would listen for incoming data, but that's a very substantial project and I wonder if there would be any use for it
This also allows for event-based scheduling of async contexts instead of launching RasterIOs in separate threads
But my use case is Express and Express-like APIs that read from GDAL data sets and my first version will simply use RasterIO in V8 AsyncWorker threads
This is slower than scheduling GDALAsyncReader contexts, but it is much simpler

**Comment by mmomtchev** [comment]
@contra, I just pushed early basic support for async reading
The current call is rather ugly:
```js
dataset.bands.get(1).pixels.readAsync(0, 0, dataset.rasterSize.x, dataset.rasterSize.y, data,
        undefined, undefined, undefined, undefined, undefined, (e, r) => {})
```

**Comment by mmomtchev** [comment]
@contra, ok I have a first version of alpha quality with async reading that satisfies my use case (reading from Express)
I will need some feedback on this, if anyone wants to try it:
https://github.com/mmomtchev/node-gdal-async
I solved the API problems in the .js layer just like the original author

**Comment by contra** [comment]
@contra Nice! Just did a quick review of the code - to keep the diff clean between the repos and make future merging easier do you want to send a PR with the changes unrelated to the async behavior?

**Comment by mmomtchev** [comment]
moved to #19

**Comment by mmomtchev** [comment]
@contra I have started the N-API migration
There is lots of automatic conversion/regexping but at the end all the `::New()` function must be almost completely rewritten
@weagle08 does your offer still stand?

**Comment by weagle08** [comment]
@mmomtchev it does, my team and I actually just started looking at this this past week. we have 4 developers with about 4 months of time allotted to this. We are all novices in making native modules though so hopefully that will be enough. Our current approach is to:

1. understand build for gdal and each dependent library
2. learn how gyp works (can use current node-gdal-next repo as guide)
3. update gyp files
4. begin wrapping objects and methods using N-API
5. determine if we can create the module without including the dependency code (hopefully can use npm commands with gulp to pull dependent sources) and update gyp files for this approach

What we won't have time to do and will need help with:
1. update all unit tests
2. async functionality
3. verify functionality of module on linux and mac (as our target system is windows 10)

Where we currently are:

We currently have most dependencies building using VC++ and are including JPEG and zlib in the build for libtiff (let me know if this is not needed). We should start on understanding gyp in the next week or 2 so any pointers on how gyp works that may be confusing/lacking from their documentation is welcome.

Please check back here for questions. I know it sounds like we are re-inventing the wheel on 1 & 2 but I think it is important for my team to understand the complete build process in case we run into issues and I'd like to better understand for my own knowledge of gyp and native modules.

**Comment by mmomtchev** [comment]
@weagle08 My current estimate is that there is about a man-week of work for the N-API migration for someone who has used NAN before but is new to N-API
There is an automatic conversion tool, after I ran it, I spent one day regexping everything that the tool missed
What is left is rewriting all the `::New` functions of the JS-visible C++ objects - the so-called `ObjectWrap` - this is the major difference between NAN and N-API:
* In NAN the C++ constructor is an empty shell, the real object construction takes place in a JS-callable ``::New``, all JS-callable methods are static and when invoked, you have to extract the `this` pointer from the `ObjectWrap` out of V8 memory space
* In N-API the JS object and the C++ object are truly one object, when JS calls `new` it lands in your C++ constructor which should support calling from JS, and when a member function is called, it is called on the C++ object with `this` pointing to the C++ object

This is what needs changing. So it took me about 1 day to run the automated conversion, understand all this, and do the first 2 classes. I guess that one week is a good estimate for the rest. The branch is based on the `universal-async` branch - which has diverged a lot from the main branch. Normally this will be merged back.

My main target is Linux, but I am usually building and testing on all 3 platforms.

**Comment by weagle08** [comment]
@mmomtchev thanks for that explanation.. that will help a lot and it sounds like N-API is much improved over NAN, or I just prefer that paradigm you described anyways.

**Comment by mmomtchev** [comment]
N-API (in fact this is [node-addon-api](https://github.com/nodejs/node-addon-api) - N-API is the low level C ABI) allows for cleaner C++ code by leveraging its existing object management features instead of implementing a new object framework.

But its main strength is that it is a stable ABI, that is, binaries does not require recompilation. You can build one executable per platform (Linux, Win32, OSX) and then distribute it without regard to the user version. Currently, @contra has something like 30 different build platforms.

@weagle08 why do you want to move the dependencies out of the source tree? Most similar projects in the Node ecosystem (Node being the first one) pack their dependencies. It allows you to make an independent and fully contained release - otherwise you will need to checkout sources with different tags from different repositories - or - to publish your own source tarballs (like Debian for example).

**Comment by weagle08** [comment]
@mmomtchev it was a bullet item on the https://github.com/contra/node-gdal-next/blob/master/ROADMAP.md that I figured we could know I out by simple performing some npm preinstall steps to pull dependent sources from a tag.. did I misunderstand that bullets intent? Or would this be more complexity with little value? I figured it would make upgrades easier

**Comment by mmomtchev** [comment]
Yes, in fact, it is only now that I am reading this in detail 😄 

Async API is almost finished, it is available on <https://github.com/mmomtchev/node-gdal-async>
@contra is testing it in production, there are lots of people that regularly clone my repo so I guess some of them are using it
If no one complains during the next few weeks, it will probably get merged into the master branch

I am currently working on the N-API migration and it will probably be finished before the async API is merged - as it is a very invasive change that will have to be merged with the async API, I have based my branch on the async branch

The streaming API is difficult (even impossible for some formats) and won't happen soon, it requires lots of new code and I am afraid that the global benefit to cost factor is simply below the required minimum

Rewriting the Geometry classes is a priority item too - and not only because of the performance problems of the serialization code - depending on how long my current situation drags on - I will probably do it next

And `node-sqlite3` still does pack its dependencies - only it is in a much cleaner way, with the original source tarballs that get extracted when building - maybe this is what @contra meant?

If you want to work on the N-API migration, I can publish my current branch, but it is code that still does not compile and it is based upon an experimental branch that is still not merged itself. And alas, nothing can be done in parallel, until the N-API is merged because it impacts roughly one out of two lines of code.

**Comment by weagle08** [comment]
@mmomtchev So that will be a huge issue because I think the N-API development will completely trash the async implementation when we do it.. would it be easier for us to do the N-API upgrade and then redo the async?

**Comment by contra** [comment]
FWIW the ROADMAP file was just thrown together for me to track my thoughts on potential improvements and things to look into - should all be taken loosely. Specifically the comment under the N-API bullet was just to look at newer node native libraries for ideas on structuring the build system, since things have come a long way in the ecosystem since this project was initially developed. Our current set of gyp files have loads of stuff in them that I want to strip down to the bare bones, I think a lot of it isn't needed anymore and what is still needed needs to be documented better for maintainability. It isn't necessarily coupled with switching to N-API, but something I figured would be good to do at the same time which is why I put it under that point.

**Comment by caldwellc** [comment]
Migrating to N-API would be great. It sounds like this change could help when building against newer versions of electron. Currently unable to get node-gdal-next to build on Windows 10 for electron 10 (which uses node 12.16.3 and v8 8.5)

**Comment by mmomtchev** [comment]
I have started it, but my branch is from the `universal-async` branch
Once I get it to at least compile, I will publish it on my fork
I have other things on my hands atm, but I will probably switch back to gdal in the coming weeks

**Comment by weagle08** [comment]
We have not forgotten about this task... Just other issues can pop up and take precedence at work so just wanted to let you guys know we are still working towards converting to NAPI

**Comment by mmomtchev** [comment]
I am currently very busy with <https://github.com/mmomtchev/rlayers>, so I haven't progressed very much with gdal. Still, I think that the `universal-async` branch needs to be merged first as it is a very far-reaching evolution. I was hoping to get some feedback before merging. I see that the branch has been cloned quite a few times - so either it is working perfectly, either it is not working at all? 😄

**Comment by weagle08** [comment]
Hey guys,
@contra  @mmomtchev 
I've started work on N-API (again, finally... other work took priority and bumped this down the list) and would like opinions on a couple changes.
1. I've now created gulp tasks that run on install that pull all dependent code from their github repositories (or zip file download) specific tag version instead of carrying around the entire source in this repo... are you guys opposed to this approach?
2. Since this is a complete re-write of the original forked wrapper would there be any objection to changing the license to MIT (or another license that is more commercial friendly) instead of Apache 2.0?  We can use Apache 2.0 but MIT is just better (I've been told) for groups using it in a commercial product.

Thanks

**Comment by contra** [comment]
@weagle08 I don't have any issue changing the license but since this is a fork of another project, it is their license. I don't know enough about it to be confident we wouldn't violate their license by changing ours to MIT even if 99% of the code is different, and the 1% still remains.
